### PR TITLE
Add truncation_side option to tokenizers

### LINKED ISSUE #12909: Truncating the prefix of a sequence rather than the suffix
# 🚀 Feature request

Hi, tokenizers get `truncation` as an argument. When set to `True` the tokenizer will truncate the suffix of a sequence so it does not surpass the specified `max_length`. I'd like to have a functionality that truncates the prefix of the sequence, so the model will see the suffix of the sequence.

## Motivation

In many applications (e.g. Dialog, and QA) the most important part of the sequence is the suffix (e.g. the question after the context, or the last response of the dialog).

## Your contribution

Perhaps I'll submit a PR, but it might take me some time as I'm close to some deadlines of mine :(

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

As requested by #12909, it would be handy if one could also decide on whether to truncate sequences from the left instead of from the right. 

As we already have a `padding_side` (which can be either left/right), it makes sense to also add a `truncation_side` (which by default is set to `"right"`, but users can initialize a tokenizer with `truncation_side` set to `"left"`).

The test could possibly be improved (for which I'd like to get some help).

Also requesting review from @patrickvonplaten since I've also added the option in `feature_extraction_sequence_utils.py`. 

Regarding the fast tokenizers, I see `padding_side` is used [here](https://github.com/huggingface/transformers/blob/12e02e339f6d19218b36a30883188ea0254bc7e7/src/transformers/tokenization_utils_fast.py#L362). Should I define something similar for `truncation_side`?

Fixes #12909

### CODE REVIEW

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -1463,6 +1469,12 @@ def __init__(self, **kwargs):
             "right",
             "left",
         ], f"Padding side should be selected between 'right' and 'left', current value: {self.padding_side}"
+        # Truncation side is right by default and overridden in subclasses. If specified in the kwargs, it is changed.
+        self.truncation_side = kwargs.pop("truncation_side", self.truncation_side)
+        assert self.truncation_side in [
+            "right",
+            "left",
+        ], f"Truncation side should be selected between 'right' and 'left', current value: {self.truncation_side}"
```
Comment (sgugger) [inline src/transformers/tokenization_utils_base.py:1477]:
```suggestion
        if self.truncation_side not in ["right", "left"]:
            raise ValueError(f"Truncation side should be selected between 'right' and 'left', current value: {self.truncation_side}")
```

File: tests/test_tokenization_common.py [hunk 2]
```diff
@@ -1502,6 +1502,29 @@ def test_encode_plus_with_padding(self):
...
+
+                # LEFT TRUNCATION
+                tokenizer.truncation_side = "left"
+                truncated_sequence_left = tokenizer.encode(
+                    sequence,
+                    max_length=max_length,
+                    truncation=True,
+                )
+                truncated_sequence_left_length = len(truncated_sequence_left)
+                assert max_length == truncated_sequence_left_length
```
Comment (sgugger) [inline tests/test_tokenization_common.py:1520]:
```suggestion
                self.assertEqual(max_length, truncated_sequence_left_length)
```

File: tests/test_tokenization_common.py [hunk 3]
```diff
@@ -1502,6 +1502,29 @@ def test_encode_plus_with_padding(self):
...
+
+                # RIGHT TRUNCATION
+                tokenizer.truncation_side = "right"
+                truncated_sequence_right = tokenizer.encode(sequence, max_length=max_length, truncation=True)
+                truncated_sequence_right_length = len(truncated_sequence_right)
+                assert max_length == truncated_sequence_right_length
```
Comment (sgugger) [inline tests/test_tokenization_common.py:1526]:
```suggestion
                self.assertEqual(max_length, truncated_sequence_right_length)
```

File: src/transformers/feature_extraction_sequence_utils.py [hunk 4]
```diff
@@ -342,9 +345,20 @@ def _truncate(
...
+                if "attention_mask" in processed_features:
+                    processed_features["attention_mask"] = processed_features["attention_mask"][:max_length]
+            elif self.truncation_side == "left":
+                processed_features[self.model_input_names[0]] = processed_features[self.model_input_names[0]][
+                    max_length:
+                ]
+                if "attention_mask" in processed_features:
+                    processed_features["attention_mask"] = processed_features["attention_mask"][max_length:]
+            else:
+                raise ValueError("invalid truncation strategy:" + str(self.truncation_side))
```
Comment (sgugger) [inline src/transformers/feature_extraction_sequence_utils.py:361]:
```suggestion
                raise ValueError(f"invalid truncation strategy: {self.truncation_side}")
```

**Review by sgugger — APPROVED** [review]
Thanks for your PR! I left some nits.
I'm not entirely sure this will work without the fast tokenizers supporting this on the Tokenizers side too, however.

**Review by LysandreJik — APPROVED** [review]
The added test is failing for TAPAS:
`FAILED tests/test_tokenization_tapas.py::TapasTokenizationTest::test_right_and_left_truncation`

Pinging @SaulLu regarding the fast tokenizer question.

Other than that, LGTM

File: tests/test_tokenization_common.py [hunk 5]
```diff
@@ -1502,6 +1502,29 @@ def test_encode_plus_with_padding(self):
...
+                    truncation=True,
+                )
+                truncated_sequence_left_length = len(truncated_sequence_left)
+                self.assertEqual(max_length, truncated_sequence_left_length)
+
+                # RIGHT TRUNCATION
+                tokenizer.truncation_side = "right"
+                truncated_sequence_right = tokenizer.encode(sequence, max_length=max_length, truncation=True)
+                truncated_sequence_right_length = len(truncated_sequence_right)
+                self.assertEqual(max_length, truncated_sequence_right_length)
```
Comment (SaulLu) [inline tests/test_tokenization_common.py:1526]: I propose in this test to also check that the encoding values correspond to the left or right values respectively. I propose here a solution without taking into account the special tokens, which seem to me more complicated to take into account automatically in the verification.

This test allows to observe that the new feature does not work correctly for the fast versions of tokenizers.

```suggestion
                sequence = "I wanted to go running, swimming and singing."
                max_length = 5

                tokenize_sequence = tokenizer(sequence, add_special_tokens=False)

                # LEFT TRUNCATION
                tokenizer.truncation_side = "left"
                truncated_sequence_left_without_special_tokens = tokenizer(
                    sequence, max_length=max_length, truncation=True, add_special_tokens=False
                )
                truncated_sequence_left_with_special_tokens = tokenizer(
                    sequence,
                    max_length=max_length,
                    truncation=True,
                )

                for key in tokenize_sequence.keys():
                    self.assertEqual(max_length, len(truncated_sequence_left_with_special_tokens[key]))
                    self.assertListEqual(
                        tokenize_sequence[key][-max_length:], truncated_sequence_left_without_special_tokens[key]
                    )

                # RIGHT TRUNCATION
                tokenizer.truncation_side = "right"
                truncated_sequence_right_without_special_tokens = tokenizer(
                    sequence, max_length=max_length, truncation=True, add_special_tokens=False
                )
                truncated_sequence_right_with_special_tokens = tokenizer(
                    sequence,
                    max_length=max_length,
                    truncation=True,
                )

                for key in tokenize_sequence.keys():
                    self.assertEqual(max_length, len(truncated_sequence_right_with_special_tokens[key]))
                    self.assertListEqual(
                        tokenize_sequence[key][:max_length], truncated_sequence_right_without_special_tokens[key]
                    )
```

**Review by SaulLu — COMMENTED** [review]
Thank you very much @NielsRogge for working on this feature! :smile: 

Unfortunately, I agree with @sgugger : I think we need to integrate this feature also in the :hugs:  `tokenizers` library for the feature to work for fast tokenizers in :hugs: `transformers` (see the errors returned by my test extension proposal ). Indeed, fast tokenizers in their `__call__` method never call the `truncate_sequences` method.

File: src/transformers/feature_extraction_sequence_utils.py [hunk 6]
```diff
@@ -316,7 +317,9 @@ def _truncate(
         truncation: Optional[bool] = None,
     ):
         """
-        Pad inputs (on left/right and up to predefined length or max length in the batch)
+        Truncate inputs (on left/right and up to predefined length or max length in the batch)
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_sequence_utils.py:320]: Can we remove this for now from the feature extraction sequence utils? :-) It'll cause problems with the normalization at the moment, so I'd prefer to not added

File: src/transformers/feature_extraction_sequence_utils.py [hunk 7]
```diff
@@ -342,9 +345,20 @@ def _truncate(
...
-                processed_features["attention_mask"] = processed_features["attention_mask"][:max_length]
+            if self.truncation_side == "right":
+                processed_features[self.model_input_names[0]] = processed_features[self.model_input_names[0]][
+                    :max_length
+                ]
+                if "attention_mask" in processed_features:
+                    processed_features["attention_mask"] = processed_features["attention_mask"][:max_length]
+            elif self.truncation_side == "left":
+                processed_features[self.model_input_names[0]] = processed_features[self.model_input_names[0]][
+                    max_length:
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_sequence_utils.py:356]: What happens here actually if `max_length` is larger then one input? E.g. imagine we have the following inputs:

`[[1, 2, 3, 4, 5], [1, 2, 3], [1, 2]]` and we want to truncate & pad to a length of 3 => with `truncation_side == "left"` this fails since here: https://github.com/huggingface/transformers/blob/3ff2cde5ca4a2d3c622b827d9edf7e3d0b7f4fb7/src/transformers/feature_extraction_sequence_utils.py#L214 we would truncate the last list `[1, 2]` essentially to an empty list no?

File: src/transformers/feature_extraction_sequence_utils.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/transformers/feature_extraction_sequence_utils.py:320]: This PR helped me found some redundant code :-) https://github.com/huggingface/transformers/pull/13051

File: src/transformers/feature_extraction_sequence_utils.py (hunk 7 above)
Comment (patrickvonplaten) [reply src/transformers/feature_extraction_sequence_utils.py:356]: Actually, I think we should just change it to `[-max_length:]` and it should work

File: src/transformers/feature_extraction_sequence_utils.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/transformers/feature_extraction_sequence_utils.py:320]: Okey thinking a bit more about it, it should actually be fine to add this feature if the comments here: https://github.com/huggingface/transformers/pull/12913/files#r685115702 are applied & tests are added to https://github.com/huggingface/transformers/blob/3ff2cde5ca4a2d3c622b827d9edf7e3d0b7f4fb7/tests/test_sequence_feature_extraction_common.py#L361 . It could be as simple as just adding a generic parameter `truncation_side` to the test function: https://github.com/huggingface/transformers/blob/3ff2cde5ca4a2d3c622b827d9edf7e3d0b7f4fb7/tests/test_sequence_feature_extraction_common.py#L221 which is then tested for both left and right

**Review by patrickvonplaten — CHANGES_REQUESTED** [review]
In the current form the PR would lead to errors in the feature extraction utils file. Think it can easily be resolved though. Also tests would need to be added for the sequence extraction scripts

**Comment by LysandreJik** [comment]
Gentle ping @NielsRogge for when you have a chance to wrap this up

**Comment by LSinev** [comment]
It is waited for more than a year now (https://github.com/huggingface/transformers/issues/4476#issuecomment-677823688). Please implement this (even if it works out just for ordinary tokenizers), for people to use now this solution, while users of Rust tokenizers wait for the fast tokenizers solution (https://github.com/huggingface/tokenizers/issues/779).

File: src/transformers/feature_extraction_sequence_utils.py (hunk 7 above)
Comment (NielsRogge) [reply src/transformers/feature_extraction_sequence_utils.py:342]: Fixed this.

File: src/transformers/feature_extraction_sequence_utils.py [hunk 8]
```diff
@@ -304,15 +305,17 @@ def _truncate(
...
 
         Args:
-            processed_features: Dictionary of input values (`np.ndarray[float]`) / input vectors (`List[np.ndarray[float]]`) or batch of inputs values (`List[np.ndarray[int]]`) / input vectors (`List[np.ndarray[int]]`)
+            processed_features: Dictionary of input values (*np.ndarray[float]*) / input vectors (*List[np.ndarray[float]]*) or batch of inputs values (*List[np.ndarray[int]]*) / input vectors (*List[np.ndarray[int]]*)
             max_length: maximum length of the returned list and optionally padding length (see below)
             pad_to_multiple_of: (optional) Integer if set will pad the sequence to a multiple of the provided value.
                 This is especially useful to enable the use of Tensor Core on NVIDIA hardware with compute capability
                 >= 7.5 (Volta), or on TPUs which benefit from having sequence lengths be a multiple of 128.
-            truncation: (optional) Activates truncation to cut input sequences longer than `max_length` to `max_length`.
+            truncation: (optional) Activates truncation to cut input sequences longer than *max_length* to *max_length*.
```
Comment (NielsRogge) [inline src/transformers/feature_extraction_sequence_utils.py:318]: @sgugger there's a weird issue with doc-builder, after applying convert it turned all code pieces into bold text.

Comment (sgugger) [reply src/transformers/feature_extraction_sequence_utils.py:318]: That's because you had a mix of Markdown and rst in the same docstring: double backticks line 310 for `self.truncation_side` and the rest in Markdown

**Comment by NielsRogge** [comment]
Fixed per #14947.
